### ISSUE TITLE
"MergeBatchNorm" canonizer and "ParamMod" leave "_parameters" desynchronized from "__dict__" after composite context exits

### ISSUE THREAD

**Issue by lecode-official** [chr5tphr/zennit#237]
First of all: sorry for the "AI slop". But while working with Zennit I had a problem that I could not quite find the root cause of, but Claude saved my butt, so I thought it was only fitting to let them write the bug report. And quite frankly, they did a much more comprehensive, thorough and well-written report than I would have ever made and most likely could not make. Here it goes:

## Description

When a `Composite` with a `MergeBatchNorm`-based canonizer (e.g., `SequentialMergeBatchNorm`) is used in a `composite.context(model)` block, the model's internal parameter references become silently corrupted after the context exits. Specifically, for every module whose parameters were both (a) merged by the canonizer and (b) modified by a `ParamMod` during the backward pass, `module._parameters` and `module.__dict__` end up holding **different** `Parameter` objects. This causes `model.parameters()`, `model.state_dict()`, and `model.named_parameters()` to return stale parameter objects, while the forward pass (`self.weight`, `self.bias`) uses the correct ones.

This is a **silent** corruption: no error is raised, and the model appears to work normally. However, any downstream code that relies on `model.parameters()` or `model.state_dict()` — such as optimizers, model checkpointing, or parameter copying — operates on the wrong tensors.

## Affected Versions

- **Zennit**: 1.0.0 (likely all versions using the current `MergeBatchNorm` + `ParamMod` implementation)
- **PyTorch**: 2.9.1 (likely any version where `load_state_dict(assign=True)` uses `setattr`)

## Root Cause

The bug arises from an interaction between two mechanisms that modify module parameters in incompatible ways:

### 1. `MergeBatchNorm` uses `object.__setattr__` (bypasses `nn.Module.__setattr__`)

In [`canonizers.py`](https://github.com/chr5tphr/zennit/blob/master/src/zennit/canonizers.py), `MergeBatchNorm.merge_batch_norm()` modifies parameters via:

```python
object.__setattr__(module, 'weight', module.weight * scale[index])
object.__setattr__(module, 'bias', (module.bias - batch_norm.running_mean) * scale + batch_norm.bias)
```

`object.__setattr__` writes directly into `module.__dict__`, **bypassing** `nn.Module.__setattr__`. This means `module.__dict__['weight']` gets the merged tensor, but `module._parameters['weight']` still holds the original `Parameter`. This is intentional — the canonizer saves the originals in `self.linear_params` and restores them in `remove()` using the same `object.__setattr__` mechanism.

### 2. `ParamMod.__call__` uses `load_state_dict(assign=True)` (uses `nn.Module.__setattr__`)

In [`core.py`](https://github.com/chr5tphr/zennit/blob/master/src/zennit/core.py), `ParamMod.__call__` temporarily modifies parameters via:

```python
module.load_state_dict(modified_state, strict=False, assign=True)
# ... yield ...
module.load_state_dict(original_state, strict=False, assign=True)
```

With `assign=True`, `load_state_dict` calls `setattr(self, name, input_param)`, which goes through `nn.Module.__setattr__`. For `Parameter` values, this **removes** the entry from `__dict__` and writes to `_parameters` instead (see [PyTorch source](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L1968-L1990)).

### 3. The conflict

Here is the sequence of events for a Conv2d layer adjacent to a BatchNorm layer during a composite context:

| Step | Action | `__dict__['weight']` | `_parameters['weight']` |
|------|--------|---------------------|------------------------|
| 1 | Initial state | *(not present)* | `P_orig` (original Parameter) |
| 2 | Canonizer `register()` saves `P_orig`, then calls `merge_batch_norm()` which does `object.__setattr__(module, 'weight', merged_weight)` | `merged_weight` (shadows `_parameters`) | `P_orig` (unchanged) |
| 3 | During LRP backward, `ParamMod.__call__` enters: `load_state_dict(modified_state, assign=True)` → `nn.Module.__setattr__` removes `__dict__['weight']`, writes new Parameter to `_parameters` | *(removed)* | `P_modified` (new Parameter from ParamMod) |
| 4 | `ParamMod.__call__` exits: `load_state_dict(original_state, assign=True)` → writes another new Parameter to `_parameters` | *(not present)* | `P_restored` (new Parameter, **not** `P_orig`) |
| 5 | Canonizer `remove()` restores: `object.__setattr__(module, 'weight', P_orig)` | `P_orig` (correct original) | `P_restored` (stale, **not** `P_orig`) |

After step 5, the module is in an inconsistent state:

- **Forward pass**: `self.weight` resolves to `__dict__['weight']` = `P_orig` (**correct**, because `__dict__` takes priority over `__getattr__` → `_parameters` in Python's attribute resolution)
- **`model.parameters()` / `model.state_dict()`**: reads from `_parameters['weight']` = `P_restored` (**wrong**, a stale object with the same data but a different identity)

### For single-use LRP attribution (no training)
The immediate attribution results are **correct** — the corruption only affects post-context model state. If the model is discarded after attribution, there is no observable bug.

### For iterative use (training loops, federated learning, etc.)
The corruption is **catastrophic** for any workflow that uses the model after attribution:

1. **Optimizers track the wrong parameters**: After `composite.context()` exits, `model.parameters()` returns `P_restored` objects. If an optimizer is (re-)initialized with these, it tracks tensors that are **not** in the computation graph of the forward pass (which uses `P_orig` from `__dict__`). Gradient updates go to the wrong tensors, and the model effectively stops learning for all affected layers.

2. **`state_dict()` returns stale data**: `model.state_dict()` reads from `_parameters`, returning the `P_restored` tensors. If these are used for checkpointing, model aggregation (e.g., federated learning), or `copy_from()`, the saved/copied parameters do not reflect the actual model state used in the forward pass.

3. **`copy_from()` / `load_state_dict()` writes to the wrong tensors**: Copying parameters via `state_dict()` + `load_state_dict()` or similar mechanisms writes data into `_parameters` entries that are not used by the forward pass.

4. **Only affects models with BatchNorm**: Models without BatchNorm layers (e.g., LeNet-5) are unaffected because the canonizer performs no merging, so no `object.__setattr__` calls occur, and `_parameters` is never shadowed by `__dict__`.

## Minimal Reproduction

```python
import torch
from zennit.composites import EpsilonPlus
from zennit.canonizers import SequentialMergeBatchNorm

# Simple model with Conv + BatchNorm
model = torch.nn.Sequential(
    torch.nn.Conv2d(1, 8, 3, padding=1),
    torch.nn.BatchNorm2d(8),
    torch.nn.ReLU(),
    torch.nn.AdaptiveAvgPool2d(1),
    torch.nn.Flatten(),
    torch.nn.Linear(8, 10),
)
model.eval()

# Get the Conv2d module and its original weight Parameter
conv = model[0]
original_weight_param = conv._parameters['weight']

# Verify initial consistency
assert 'weight' not in conv.__dict__, "weight should NOT be in __dict__ initially"
assert conv._parameters['weight'] is original_weight_param

# Run LRP attribution inside composite context
composite = EpsilonPlus(canonizers=[SequentialMergeBatchNorm()])
input_tensor = torch.randn(1, 1, 8, 8, requires_grad=True)

with composite.context(model) as modified_model:
    output = modified_model(input_tensor)
    torch.autograd.backward(output, torch.ones_like(output))

# After context exits: check for the bug
weight_in_dict = conv.__dict__.get('weight')
weight_in_params = conv._parameters.get('weight')

print(f"weight in __dict__: {weight_in_dict is not None}")       # True (should be False)
print(f"weight is original: {weight_in_dict is original_weight_param}")  # True
print(f"_parameters is original: {weight_in_params is original_weight_param}")  # False! BUG!
print(f"_parameters is __dict__: {weight_in_params is weight_in_dict}")  # False! BUG!

# The forward pass uses __dict__['weight'] (correct), but model.parameters()
# returns _parameters['weight'] (stale). An optimizer initialized now would
# track the wrong tensors.
params_list = list(model.parameters())
assert any(p is original_weight_param for p in params_list), \
    "BUG: model.parameters() does not include the actual weight used in forward pass"
```

## Workaround

After the composite context exits, reconcile `_parameters` with `__dict__` for all modules:

```python
with composite.context(model) as modified_model:
    # ... perform attribution ...
    pass

# Fix: move correct Parameters from __dict__ back into _parameters
for module in model.modules():
    for param_name in list(module._parameters.keys()):
        if param_name in module.__dict__:
            module._parameters[param_name] = module.__dict__.pop(param_name)
```

## Suggested Fix

The root cause is that `MergeBatchNorm.merge_batch_norm()` and `MergeBatchNorm.remove()` use `object.__setattr__` to bypass `nn.Module.__setattr__`, which writes to `__dict__` instead of `_parameters`. When `ParamMod` subsequently uses `load_state_dict(assign=True)`, which goes through `nn.Module.__setattr__`, the `__dict__` entries are cleaned up and `_parameters` is updated — but with *new* Parameter objects, not the originals the canonizer saved.

A clean fix could take several forms:

1. **Make `MergeBatchNorm` use `nn.Module.__setattr__`** (or a parameter-swap mechanism) instead of `object.__setattr__`, so that all modifications go through `_parameters` consistently. This would require adjusting the save/restore logic since `nn.Module.__setattr__` enforces type constraints (e.g., requiring `Parameter` objects for existing parameter names).

2. **Make `ParamMod.__call__` use `object.__setattr__`** to be consistent with the canonizer. However, this would mean `_parameters` is never updated, which could break other code paths.

3. **Add reconciliation in `Composite.context.__exit__`** (or `Canonizer.remove`): after removing the canonizer, iterate over all affected modules and ensure `_parameters` and `__dict__` are consistent, similar to the workaround above.

## Related Issues

- #204 — "Potential issue with learning BatchNorm parameters": The user reports that models with BatchNorm have drastically different outputs in train vs. eval mode when using composite context during training. This is very likely the same underlying bug: after the composite context corrupts `_parameters`, the optimizer updates stale parameter objects, causing the actual parameters (in `__dict__`) to diverge from what the optimizer thinks it's optimizing.

- #200 — "Model inference outside of composite context": Related discussion about the interaction between composite contexts and model inference/training.

**Comment by chr5tphr** [comment]
Hey David,

Thank you for the issue! Yes, this makes sense to me. I think I used `object.__setattr__` because I did not want to trigger the torch.nn.Module mechanic for some reason, but I currently cannot remember why.
Since Pytorch also changed a lot, it may also be that, whatever concern I used to have, does not exist anymore.
I think there was something similar going on with `zennit.core.with_params` (which is now `zennit.core.ParamMod`),  which  I changed to use `torch.nn.Module.load_state_dict(..., assign=True)`. I think it would be much cleaner to implement it the same way for the canonizers.
